apps/ejabberd/src/mod_mam_utils.erl

```diff
@@ -641,19 +647,25 @@ find_field([], Name) ->
 field_to_value(FieldEl) ->
     xml:get_path_s(FieldEl, [{elem, <<"value">>}, cdata]).
 
--spec message_form(binary()) -> jlib:xmlel().
-message_form(MamNs) ->
+-spec message_form(Mod :: mod_mam | mod_mam_muc, Host :: ejabberd:lserver(), binary()) ->
+    jlib:xmlel().
+message_form(Module, Host, MamNs) ->
     SubEl = #xmlel{name = <<"x">>,
-           attrs = [{<<"xmlns">>, <<"jabber:x:data">>},
-                    {<<"type">>, <<"form">>}],
-           children = message_form_fields(MamNs)},
+                   attrs = [{<<"xmlns">>, <<"jabber:x:data">>},
+                            {<<"type">>, <<"form">>}],
+                   children = message_form_fields(Module, Host, MamNs)},
     result_query(SubEl, MamNs).
 
-message_form_fields(MamNs) ->
+message_form_fields(Mod, Host, MamNs) ->
+    TextSearch =
+        case Mod:has_full_text_search(Host) of
+            true -> [form_field(<<"text-single">>, <<"full-text-search">>)];
+            false -> []
+        end,
     [form_type_field(MamNs),
      form_field(<<"jid-single">>, <<"with">>),
      form_field(<<"text-single">>, <<"start">>),
-     form_field(<<"text-single">>, <<"end">>)].
+     form_field(<<"text-single">>, <<"end">>)] ++ TextSearch.
```

Comment:
I know it's not the same in general, but isn't it more idiomatic (and marginally faster) in this case?